File: core/src/main/java/com/devexperts/dxlab/lincheck/annotations/Operation.java
package com.devexperts.dxlab.lincheck.annotations;

/*
 * #%L
 * core
 * %%
 * Copyright (C) 2015 - 2017 Devexperts, LLC
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Mark your method with this annotation to use it in concurrent testing as an operation.
 * <p>
 * Here are several parameters:
 * <ul>
 *     <li> {@code params} - binds by name arguments of the operation
 *          with {@link Param} definitions on the test class
 *     </li>
 *     <li> {@code runOnce} - set it to {@code true} if you want this operation
 *     to be called at most once during the test invocation; {@code false} by default.
 *     </li>
 * </ul>
 * See also {@link HandleExceptionAsResult}
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface Operation {
    String[] params() default {};
    boolean runOnce() default false;
    String group() default "";
}

File: src/jvm/main/org/jetbrains/lincheck/datastructures/Operation.kt
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2023 JetBrains s.r.o.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
 * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package org.jetbrains.lincheck.datastructures

/**
 * Mark your method with this annotation to use it in concurrent testing as an operation.
 */
@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
annotation class Operation(
    /**
     * Binds the arguments of this operation with the specified [parameter configurations][Param]
     * by their [Param.name] names.
     */
    val params: Array<String> = [],

    /**
     * Set it to `true` if you this operation should be called
     * at most once during the test invocation; `false` by default.
     */
    val runOnce: Boolean = false,

    /**
     * Operations of the same group are never executed concurrently.
     */
    val nonParallelGroup: String = "",

    /**
     * Specifies whether the operation can be cancelled if it suspends,
     * see `CancellableContinuation.cancel`; `true` by default.
     */
    val cancellableOnSuspension: Boolean = true,

    /**
     * Specifies whether this operation is blocking.
     * This way, if the test checks for a non-blocking progress guarantee,
     * **lincheck** will not fail the test if a hang is detected on
     * a running operation with this `blocking` marker.
     */
    val blocking: Boolean = false,

    /**
     * Specifies whether this cancellable operation supports
     * prompt cancellation, `false` by default. This parameter
     * is ignored if [.cancellableOnSuspension] is `false`.
     */
    val promptCancellation: Boolean = false,
)
